Add support for Tasmota camera

anishsane opened the pull request:
## Proposed change
ESP32-cam (and other esp32 variants with camera support) have been around for a while and the tasmota32-webcam firmware has been stable.
The HA integration supports the other entity types, but the camera support is not available. This PR adds the missing support for camera.
While esphome integration supports esp32-cam, tasmota firmware has many other well-documented features that can be controlled by mqtt.


## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This change is dependent on the [hatasmota change](https://github.com/emontnemery/hatasmota/pull/353). One test added fails without that PR.
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/36939

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

anishsane committed:
Add support for Tasmota camera

Enables the support for Tasmota webcam.

ESP32-cam (and other esp32 variants with camera support) have been around for a while and the tasmota32-webcam firmware has been stable.
The HA integration supports the other entity types, but the camera support is not available. This PR adds the missing support for camera.
While esphome integration supports esp32-cam, tasmota firmware has many other well-documented features that can be controlled by mqtt.

anishsane commented:
I accidentally renamed the branch used for creating the [original PR](https://github.com/home-assistant/core/pull/135706) while trying to rebase to TOT. Thus, the original PR got closed.
So, creating a fresh PR instead. Apologies for the noise.

joostlek commented on homeassistant/components/tasmota/camera.py:
```diff
@@ -0,0 +1,110 @@
+
+    """Representation of a Tasmota Camera."""
+
+    def __init__(self, **kwds: Any) -> None:
+        """Initialize a MJPEG camera."""
+        super().__init__(
+            **kwds,
+        )
+        Camera.__init__(self)
```
```suggestion
    def __init__(self, **kwargs: Any) -> None:
        """Initialize a MJPEG camera."""
        super().__init__(**kwargs)
        Camera.__init__(self)
```

joostlek commented on homeassistant/components/tasmota/camera.py:
```diff
@@ -0,0 +1,110 @@
+                return await response.read()
+
+        except TimeoutError:
+            _LOGGER.error("Timeout getting camera image from %s", self.name)
+
+        except aiohttp.ClientError as err:
+            _LOGGER.error("Error getting new camera image from %s: %s", self.name, err)
```
Instead we should raise `HomeAssistantError`, which is propagated to the frontend and the logs

anishsane replied:
Updated.

anishsane committed:
Tasmota camera: fix linter errors

emontnemery commented on homeassistant/components/tasmota/camera.py:
```diff
@@ -0,0 +1,113 @@
+        try:
+            async with asyncio.timeout(TIMEOUT):
+                response = await self._tasmota_entity.get_still_image_stream(websession)
+
```
```suggestion
```

anishsane committed:
Update homeassistant/components/tasmota/camera.py  from PR suggestions

Co-authored-by: Erik Montnemery <erik@montnemery.com>

anishsane committed:
Add tests for single frame and stream capture with tasmota camera.

This commit adds test_camera_single_frame function to test a single
frame capture and test_camera_stream function to test stream
frame capture with mocked response stream.

anishsane commented:
> @anishsane I've merged your PR on hatasmota fixing the discovery message handling and opened a PR on home assistant core bumping the library. However, how come the tests in this PR related to discovery are not failing? For example the test `test_discovery_removal_camera`?

It is failing the `test_discovery_removal_camera` test:
https://github.com/home-assistant/core/actions/runs/17483571280/job/49961689328
It will be fixed when the home assistant will start using the updated hatasmota with the above PR merged.

emontnemery reviewed (approved):
LGTM, thanks a lot @anishsane 👍 👍 

Other PRs on Tasmota would be most welcome, we can chat on Discord if you want.
